logaggregator/main.go

```diff
@@ -28,320 +22,40 @@ func main() {
-	id string,
-	n int,
-	filter Filter,
-	done <-chan struct{},
-) <-chan *rfc5424.Message {
-	msgc := make(chan *rfc5424.Message)
-	go func() {
-		buf := a.getOrInitializeBuffer(id)
-
-		var messages []*rfc5424.Message
-		var subc <-chan *rfc5424.Message
-		var cancel func()
-
-		if (filter != nil && n != 0) || n < 0 {
-			messages, subc, cancel = buf.ReadAllAndSubscribe()
+			snapshot.Close()
 		} else {
-			messages, subc, cancel = buf.ReadLastNAndSubscribe(n)
-		}
-		if filter != nil {
-			messages = filter.Filter(messages)
-			if n > 0 && len(messages) > n {
-				messages = messages[len(messages)-n:]
-			}
-		}
-		defer cancel()
-		defer close(msgc)
-
-		// range over messages, watch done
-		for _, msg := range messages {
-			select {
-			case <-done:
-				return
-			case msgc <- msg:
-			}
-		}
-
-		// select on subc, done, and cancel if done
-		for {
-			select {
-			case msg := <-subc:
-				if msgc == nil { // subc was closed
-					return
-				}
-				if !filter.Match(msg) {
-					continue // skip this message if it doesn't match filters
-				}
-				select {
-				case msgc <- msg:
-				case <-done:
-					return
-				}
-			case <-done:
-				return
-			}
+			log15.Error("error getting snapshot from leader", "error", err)
 		}
-	}()
-	return msgc
-}
-
-func (a *Aggregator) accept() {
-	defer a.listener.Close()
-
-	for {
-		select {
-		case <-a.shutdown:
-			return
-		default:
-		}
-		conn, err := a.listener.Accept()
-		if err != nil {
-			continue
-		}
-
-		a.producerwg.Add(1)
-		go func() {
-			defer a.producerwg.Done()
-			a.readLogsFromConn(conn)
-		}()
+	} else {
+		log15.Info("error finding leader for snapshot", "error", err)
 	}
-}
-
-// testing hook:
-var afterMessage func()
-
-func (a *Aggregator) getBuffer(id string) *ring.Buffer {
-	a.bmu.Lock()
-	defer a.bmu.Unlock()
-
-	buf, _ := a.buffers[id]
-	return buf
-}
-
-func (a *Aggregator) getOrInitializeBuffer(id string) *ring.Buffer {
-	a.bmu.Lock()
-	defer a.bmu.Unlock()
-
-	if buf, ok := a.buffers[id]; ok {
-		return buf
-	}
-	buf := ring.NewBuffer()
-	a.buffers[id] = buf
-	return buf
-}
-
-func (a *Aggregator) readLogsFromConn(conn net.Conn) {
-	defer conn.Close()
-
-	connDone := make(chan struct{})
-	defer close(connDone)
 
-	go func() {
-		select {
-		case <-connDone:
-		case <-a.shutdown:
-			conn.Close()
-		}
-	}()
-
-	s := bufio.NewScanner(conn)
-	s.Split(rfc6587.Split)
-	for s.Scan() {
-		msgBytes := s.Bytes()
-		// slice in msgBytes could get modified on next Scan(), need to copy it
-		msgCopy := make([]byte, len(msgBytes))
-		copy(msgCopy, msgBytes)
-
-		msg, err := rfc5424.Parse(msgCopy)
-		if err != nil {
-			log15.Error("rfc5424 parse error", "err", err)
-		} else {
-			a.msgc <- msg
-		}
-	}
-}
-
-func (a *Aggregator) run() {
-	for msg := range a.msgc {
-		a.getOrInitializeBuffer(string(msg.AppName)).Add(msg)
-		if afterMessage != nil {
-			afterMessage()
-		}
+	if err := srv.Start(); err != nil {
+		shutdown.Fatal(err)
 	}
+	<-make(chan struct{})
```

Comment:
Could rename `Server.Start` to `Server.Serve` which blocks (rather than starting a goroutine and then blocking on this channel)